docs: Add v0.18 migration (#1599)
tendermint/starport: docs/migration/index.md

---
order: 1
parent:
  title: Migration
  order: 3

> description: Required changes when upgrading from Starport v0.17.3 to Starport v0.18.

Reviewer: ```suggestion
description: For chains that were scaffolded with Starport versions lower than v0.18, changes are required to use Starport v0.18. 
```

---

> 

Reviewer: Let's define why we need these changes, how do these changes look?

Author: We need these modifications because Cosmos SDK v0.44 changed a lot how things work.

> The default template in Cosmos SDK versions lower than v0.44 are not compatible with Starport v0.18.

Reviewer: ```suggestion
Cosmos SDK v0.44 introduced changes that are not compatible with chains that were scaffolded with Starport versions lower than v0.18. 

**Important:** After upgrading from Starport v0.17.3 to Starport v0.18, you must update the default blockchain template to use blockchains that were scaffolded with earlier versions. 

```

Changes are required when upgrade from Starport v0.17.3 to Starport v0.18.0 and later. 

> To update the default template so your Starport installation is compatible with Cosmos SDK v0.44, make these changes to the `go.mod` file after you upgrade to Starport v0.18.

Reviewer: when are these changes required? after the upgrade and when Starport is started?

Author: These changes are needed, if you have a chain scaffolded with a pre-v0.18 version of Starport and you want to use v0.18 with it.

> Changes are required when upgrade from Starport v0.17.3 to Starport v0.18.0 and later. 
> 
> To update the default template so your Starport installation is compatible with Cosmos SDK v0.44, make these changes to the blockchain template after you upgrade to Starport v0.18.

Reviewer: ```suggestion
```

# v0.18

## Blockchain

> 

Reviewer: ```suggestion
For each file listed, make the required changes to the source code of the blockchain template. 

```

---
order: 1
parent:
  title: Migration
  order: 3
description: For chains that were scaffolded with Starport versions lower than v0.18, changes are required to use Starport v0.18. 
---

Cosmos SDK v0.44 introduced changes that are not compatible with chains that were scaffolded with Starport versions lower than v0.18. 

**Important:** After upgrading from Starport v0.17.3 to Starport v0.18, you must update the default blockchain template to use blockchains that were scaffolded with earlier versions. 

# v0.18

## Blockchain

> For each file listed, make the required changes to the source code of the blockchain template. 

Reviewer: ```suggestion

For each file listed, make the required changes to the source code of the blockchain template. 
```

    minttypes.ModuleName,
    //...
    // Add feegrant.ModuleName,
    feegrant.ModuleName,
  )

  // Add app.appCodec as an argument to module.NewConfigurator:
  app.mm.RegisterServices(module.NewConfigurator(app.appCodec, app.MsgServiceRouter(), app.GRPCQueryRouter()))
  
  // Replace:
  // app.SetAnteHandler(
  // 	ante.NewAnteHandler(
  // 		app.AccountKeeper, app.BankKeeper, ante.DefaultSigVerificationGasConsumer,
  // 		encodingConfig.TxConfig.SignModeHandler(),
  // 	),
  // )

  // With the following:
  anteHandler, err := ante.NewAnteHandler(
    ante.HandlerOptions{
      AccountKeeper:   app.AccountKeeper,
      BankKeeper:      app.BankKeeper,
      SignModeHandler: encodingConfig.TxConfig.SignModeHandler(),
      FeegrantKeeper:  app.FeeGrantKeeper,
      SigGasConsumer:  ante.DefaultSigVerificationGasConsumer,
    },
  )
  if err != nil {
    panic(err)
  }
  app.SetAnteHandler(anteHandler)

  // Remove the following:
  // ctx := app.BaseApp.NewUncachedContext(true, tmproto.Header{})
  // app.CapabilityKeeper.InitializeAndSeal(ctx)
}

func (app *App) InitChainer(ctx sdk.Context, req abci.RequestInitChain) abci.ResponseInitChain {
  var genesisState GenesisState

>   // Replace tmjson with json

Reviewer 2: ```suggestion
```

I should have specified this in my previous suggestion. It was `tmjson` before, we changed this to using `json` package initially but it was a mistake, this must remain `tmjson` and therefore there are no change in this part

Author: Got it, thanks.

      FeegrantKeeper:  app.FeeGrantKeeper,
      SigGasConsumer:  ante.DefaultSigVerificationGasConsumer,
    },
  )
  if err != nil {
    panic(err)
  }
  app.SetAnteHandler(anteHandler)

  // Remove the following:
  // ctx := app.BaseApp.NewUncachedContext(true, tmproto.Header{})
  // app.CapabilityKeeper.InitializeAndSeal(ctx)
}

func (app *App) InitChainer(ctx sdk.Context, req abci.RequestInitChain) abci.ResponseInitChain {
  var genesisState GenesisState
  if err := tmjson.Unmarshal(req.AppStateBytes, &genesisState); err != nil {
    panic(err)
  }
  // Add the following:
  app.UpgradeKeeper.SetModuleVersionMap(ctx, app.mm.GetVersionMap())
  return app.mm.InitGenesis(ctx, app.appCodec, genesisState)
}

// Replace Marshaler with Codec
func (app *App) AppCodec() codec.Codec {
  return app.appCodec
}

// Replace BinaryMarshaler with BinaryCodec
func initParamsKeeper(appCodec codec.BinaryCodec, legacyAmino *codec.LegacyAmino, key, tkey sdk.StoreKey) paramskeeper.Keeper {
  //...
}
```

### app/genesis.go

```go
// Replace codec.JSONMarshaler with codec.JSONCodec
func NewDefaultGenesisState(cdc codec.JSONCodec) GenesisState {
  //...
}
```

> ### testutil/keeper/mars.go

Reviewer 3: If it's an IBC module the code is different

```go
package keeper

import (
	"testing"

	"github.com/cosmonaut/test/x/mars/keeper"
	"github.com/cosmonaut/test/x/mars/types"
	"github.com/cosmos/cosmos-sdk/codec"
	codectypes "github.com/cosmos/cosmos-sdk/codec/types"
	"github.com/cosmos/cosmos-sdk/store"
	storetypes "github.com/cosmos/cosmos-sdk/store/types"
	sdk "github.com/cosmos/cosmos-sdk/types"
	capabilitykeeper "github.com/cosmos/cosmos-sdk/x/capability/keeper"
	typesparams "github.com/cosmos/cosmos-sdk/x/params/types"
	ibckeeper "github.com/cosmos/ibc-go/modules/core/keeper"
	"github.com/stretchr/testify/require"
	"github.com/tendermint/tendermint/libs/log"
	tmproto "github.com/tendermint/tendermint/proto/tendermint/types"
	tmdb "github.com/tendermint/tm-db"
)

func MarsKeeper(t testing.TB) (*keeper.Keeper, sdk.Context) {
	logger := log.NewNopLogger()

	storeKey := sdk.NewKVStoreKey(types.StoreKey)
	memStoreKey := storetypes.NewMemoryStoreKey(types.MemStoreKey)

	db := tmdb.NewMemDB()
	stateStore := store.NewCommitMultiStore(db)
	stateStore.MountStoreWithDB(storeKey, sdk.StoreTypeIAVL, db)
	stateStore.MountStoreWithDB(memStoreKey, sdk.StoreTypeMemory, nil)
	require.NoError(t, stateStore.LoadLatestVersion())

	registry := codectypes.NewInterfaceRegistry()
	appCodec := codec.NewProtoCodec(registry)
	capabilityKeeper := capabilitykeeper.NewKeeper(appCodec, storeKey, memStoreKey)

	amino := codec.NewLegacyAmino()
	ss := typesparams.NewSubspace(appCodec,
		amino,
		storeKey,
		memStoreKey,
		"MarsSubSpace",
	)
	IBCKeeper := ibckeeper.NewKeeper(
		appCodec,
		storeKey,
		ss,
		nil,
		nil,
		capabilityKeeper.ScopeToModule("MarsIBCKeeper"),
	)

	k := keeper.NewKeeper(
		codec.NewProtoCodec(registry),
		storeKey,
		memStoreKey,
		IBCKeeper.ChannelKeeper,
		&IBCKeeper.PortKeeper,
		capabilityKeeper.ScopeToModule("MarsScopedKeeper"),
	)

	ctx := sdk.NewContext(stateStore, tmproto.Header{}, false, logger)
	return k, ctx
}

```

Author: Added!

---
order: 1
title: v0.18
parent:
  title: Migration
  order: 3
description: For chains that were scaffolded with Starport versions lower than v0.18, changes are required to use Starport v0.18. 
---

# Upgrading a Blockchain to use Starport v0.18

> Starport v0.18 comes with Cosmos SDK v0.44. This version of the SDK introduced changes that are not compatible with chains that were scaffolded with Starport versions lower than v0.18.

Reviewer: ```suggestion
Starport v0.18 comes with Cosmos SDK v0.44. This version of Cosmos SDK introduced changes that are not compatible with chains that were scaffolded with Starport versions lower than v0.18.
```

**Important:** After upgrading from Starport v0.17.3 to Starport v0.18, you must update the default blockchain template to use blockchains that were scaffolded with earlier versions.

> These instructions are written for a blockchain scaffolded with the following command:

Reviewer: ```suggestion
These instructions are written for a blockchain that was scaffolded with the following command:
```

```
starport scaffold chain github.com/cosmonaut/mars
```

> If you used a different module path, replace `cosmonaut` and `mars` with correct values accordingly.

Reviewer: ```suggestion
If you used a different module path, replace `cosmonaut` and `mars` with the correct values for your blockchain.
```

  ibchost "github.com/cosmos/ibc-go/modules/core/24-host"
  ibckeeper "github.com/cosmos/ibc-go/modules/core/keeper"
  
  // Remove the following packages:
  // transfer "github.com/cosmos/cosmos-sdk/x/ibc/applications/transfer"
  // ibctransferkeeper "github.com/cosmos/cosmos-sdk/x/ibc/applications/transfer/keeper"
  // ibctransfertypes "github.com/cosmos/cosmos-sdk/x/ibc/applications/transfer/types"
  // ibc "github.com/cosmos/cosmos-sdk/x/ibc/core"
  // ibcclient "github.com/cosmos/cosmos-sdk/x/ibc/core/02-client"
  // porttypes "github.com/cosmos/cosmos-sdk/x/ibc/core/05-port/types"
  // ibchost "github.com/cosmos/cosmos-sdk/x/ibc/core/24-host"
  // ibckeeper "github.com/cosmos/cosmos-sdk/x/ibc/core/keeper"
)

var (
  //...
  ModuleBasics = module.NewBasicManager(
    //...
    slashing.AppModuleBasic{},
    // Add feegrantmodule.AppModuleBasic{},
    feegrantmodule.AppModuleBasic{}, // <--
    ibc.AppModuleBasic{},
    //...
  )
  //...
)

type App struct {
  //...
  // Replace codec.Marshaler with codec.Codec
  appCodec          codec.Codec // <--
  // Add FeeGrantKeeper
  FeeGrantKeeper   feegrantkeeper.Keeper // <--
}

> ``go

Reviewer 4: This is wrong. Either remove it, have an ending statement for the last part. Also, 3 instead of two `

Author: If you meant just the double-tick go part here, then ok.

Reviewer 4: Yes

func New(...) {
  //bApp.SetAppVersion(version.Version)
  bApp.SetVersion(version.Version) // <--

  keys := sdk.NewKVStoreKeys(
    //...
    upgradetypes.StoreKey,
    // Add feegrant.StoreKey
    feegrant.StoreKey, // <--
    evidencetypes.StoreKey,
    //...
  )

  app.FeeGrantKeeper = feegrantkeeper.NewKeeper(appCodec, keys[feegrant.StoreKey], app.AccountKeeper)  // <--
  // Add app.BaseApp as the last argument to upgradekeeper.NewKeeper
  app.UpgradeKeeper = upgradekeeper.NewKeeper(skipUpgradeHeights, keys[upgradetypes.StoreKey], appCodec, homePath, app.BaseApp)
  
  app.IBCKeeper = ibckeeper.NewKeeper(
    // Add app.UpgradeKeeper
    appCodec, keys[ibchost.StoreKey], app.GetSubspace(ibchost.ModuleName), app.StakingKeeper, app.UpgradeKeeper, scopedIBCKeeper,
  )

  govRouter.AddRoute(govtypes.RouterKey, govtypes.ProposalHandler).
    //...

>     // Replace NewClientUpdateProposalHandler with NewClientUpdateProposalHandler

Reviewer 4: Those look identical

Reviewer 4: Should be `NewClientProposalHandler` after the upgrade

    AddRoute(ibchost.RouterKey, ibcclient.NewClientUpdateProposalHandler(app.IBCKeeper.ClientKeeper))

  // Replace porttypes with ibcporttypes
  ibcRouter := ibcporttypes.NewRouter()

  app.mm.SetOrderBeginBlockers(
    upgradetypes.ModuleName,
    // Add capabilitytypes.ModuleName,
    capabilitytypes.ModuleName,
    minttypes.ModuleName,
    //...
    // Add feegrant.ModuleName,
    feegrant.ModuleName,
  )

  app.mm.RegisterServices(module.NewConfigurator(
    // Add app.appCodec

>     app.appCodec,

Reviewer 4: Could you add the whole RegisterServices here? 

Does this look right?
```
app.mm.RegisterServices(module.NewConfigurator(
		app.MsgServiceRouter(), 
		app.GRPCQueryRouter(), 
		app.appCodec,
	))
```

Author: Done!

    //...
  )
  
  // Replace:
  // app.SetAnteHandler(
  // 	ante.NewAnteHandler(
  // 		app.AccountKeeper, app.BankKeeper, ante.DefaultSigVerificationGasConsumer,
  // 		encodingConfig.TxConfig.SignModeHandler(),
  // 	),
  // )

  // With the following:
  anteHandler, err := ante.NewAnteHandler(
    ante.HandlerOptions{
      AccountKeeper:   app.AccountKeeper,
      BankKeeper:      app.BankKeeper,
      SignModeHandler: encodingConfig.TxConfig.SignModeHandler(),
      FeegrantKeeper:  app.FeeGrantKeeper,
      SigGasConsumer:  ante.DefaultSigVerificationGasConsumer,
    },
  )
  if err != nil {
    panic(err)
  }
  app.SetAnteHandler(anteHandler)

  // Remove the following:
  // ctx := app.BaseApp.NewUncachedContext(true, tmproto.Header{})
  // app.CapabilityKeeper.InitializeAndSeal(ctx)
}

func (app *App) InitChainer(ctx sdk.Context, req abci.RequestInitChain) abci.ResponseInitChain {
  var genesisState GenesisState
  // Replace tmjson with json

>   if err := json.Unmarshal(req.AppStateBytes, &genesisState); err != nil {

Reviewer 2: ```suggestion
  if err := tmjson.Unmarshal(req.AppStateBytes, &genesisState); err != nil {
```

  registry := codectypes.NewInterfaceRegistry()
  k := keeper.NewKeeper(
    codec.NewProtoCodec(registry),
    storeKey,
    memStoreKey,
  )

  ctx := sdk.NewContext(stateStore, tmproto.Header{}, false, log.NewNopLogger())
  return k, ctx
}
```

### testutil/network/network.go

```go
func DefaultConfig() network.Config {
  //...
  return network.Config{
    //...
    // Add sdk.DefaultPowerReduction
    AccountTokens:   sdk.TokensFromConsensusPower(1000, sdk.DefaultPowerReduction),
    StakingTokens:   sdk.TokensFromConsensusPower(500, sdk.DefaultPowerReduction),
    BondedTokens:    sdk.TokensFromConsensusPower(100, sdk.DefaultPowerReduction),
    //...
  }
}
```

### testutil/sample/sample.go

Add the following code:

```go
package sample

import (
  "github.com/cosmos/cosmos-sdk/crypto/keys/ed25519"
  sdk "github.com/cosmos/cosmos-sdk/types"
)

// AccAddress returns a sample account address
func AccAddress() string {
  pk := ed25519.GenPrivKey().PubKey()
  addr := pk.Address()
  return sdk.AccAddress(addr).String()
}
```

### go.mod

```
module github.com/cosmonaut/mars

go 1.16

require (
	github.com/cosmos/cosmos-sdk v0.44.0
	github.com/cosmos/ibc-go v1.2.0

> 	github.com/gogo/protobuf v1.3.3

Reviewer 3: if the user uses Bandchain oracles, he needs to update the package (`github.com/bandprotocol/bandchain-packet`) version to `v0.0.2`

Author: I've used the `go.mod` scaffolded by default with the latest `develop`. Should we also change the default `go.mod`?

Reviewer 3: We only add the `go.mod` if we scaffold an oracle component, like the wasm integration. Maybe we should add a note: "if you are using Banchain Oracle integration, you should update to version `v0.0.2`"

  cdc codec.BinaryCodec,
  //...
) *Keeper {
  // ...
}
```

### x/mars/keeper/msg_server_test.go

```go
package keeper_test

import (
  //...
  // Add the following:
  keepertest "github.com/cosmonaut/mars/testutil/keeper"
  "github.com/cosmonaut/mars/x/mars/keeper"
)

func setupMsgServer(t testing.TB) (types.MsgServer, context.Context) {
  // Replace
  // keeper, ctx := setupKeeper(t)
  // return NewMsgServerImpl(*keeper), sdk.WrapSDKContext(ctx)

  // With the following:
  k, ctx := keepertest.MarsKeeper(t)
  return keeper.NewMsgServerImpl(*k), sdk.WrapSDKContext(ctx)
}
```

### x/mars/module.go

```go
type AppModuleBasic struct {
  // Replace Marshaler with BinaryCodec
  cdc codec.BinaryCodec
}

// Replace Marshaler with BinaryCodec
func NewAppModuleBasic(cdc codec.BinaryCodec) AppModuleBasic {
  return AppModuleBasic{cdc: cdc}
}

// Replace JSONMarshaler with JSONCodec
func (AppModuleBasic) DefaultGenesis(cdc codec.JSONCodec) json.RawMessage {
  return cdc.MustMarshalJSON(types.DefaultGenesis())
}

// Replace JSONMarshaler with JSONCodec
func (AppModuleBasic) ValidateGenesis(cdc codec.JSONCodec, config client.TxEncodingConfig, bz json.RawMessage) error {
  //...
}

> // Replace JSONMarshaler with JSONCodec

Reviewer 4: ```suggestion
// Replace codec.Marshaller with codec.Codec
```

parent:
  title: Migration
  order: 3
description: Required changes when upgrading from Starport v0.17.3 to Starport v0.18.
---

The default template in Cosmos SDK versions lower than v0.44 are not compatible with Starport v0.18.

Changes are required when upgrade from Starport v0.17.3 to Starport v0.18.0 and later. 

> To update the default template so your Starport installation is compatible with Cosmos SDK v0.44, make these changes to the `go.mod` file after you upgrade to Starport v0.18.

Author: ```suggestion
To update the default template so your Starport installation is compatible with Cosmos SDK v0.44, make these changes to the blockchain template after you upgrade to Starport v0.18.
```
The changes are made to the source code of the blockchain, not just `go.mod`.
